r-simonelli/demos-lifecycle (#750)
ros2/demos: lifecycle/README.rst

Introduction
------------

What Is This?
^^^^^^^^^^^^^
ROS 2 introduces the concept of managed nodes, also called ``LifecycleNode``\ s.
In the following tutorial, we explain the purpose of these nodes, what makes them different from regular nodes and how they comply to a lifecycle management.

Basic Concept and Nomenclature
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Managed nodes contain a state machine with a set of predefined states.
These states can be changed by invoking a transition, passing a specific transition ID which indicates the succeeding consecutive state.
The state machine is implemented as described at the `Managed nodes design page <http://design.ros2.org/articles/node_lifecycle.html>`__.

> The implementation differentiates between primary states and transition states.

Reviewer: ```suggestion
The implementation differentiates between ``Primary States`` and ``Transition States``.
```

Primary states are supposed to be steady states in which any node can perform their respective task.
Transition states are used to indicate whether a transition between two primary states is considered successful or not.

> In the end, any managed node can be in one of the following states:

Reviewer: ```suggestion
 - ``Primary states`` are supposed to be steady states in which any node can perform their respective task.
 - ``Transition states`` are used to indicate whether a transition between two primary states is considered successful or not.
In the end, any managed node can be in one of the following states:
```

Primary States (steady states):

* unconfigured
* inactive
* active

> * finalized

Reviewer 2: Good point, this should be `finalized` 👍

* ``on_shutdown()`` (in the ``shuttingDown`` state): In this case does the same as ``on_cleanup()``, in general it would for example release system resources.

In summary, the node only performs its main functionality (publishing) while in the ``active`` state.
The transitions and their corresponding callbacks ensure that resources are correctly initialized,
activated, deactivated, and cleaned up as the node moves between lifecycle states.

This demo shows a simple talker/listener pair of nodes.
However, imagine a more realistic scenario with attached hardware which may have a rather long booting phase, i.e. a laser or camera.
In that case, one could initialize the device driver in the ``configuring`` transition state,
start and stop the publishing of the device’s data in the ``activating`` and ``deactivating`` transition states,
and only shut down the device completely in the ``cleaningUp`` or ``shuttingdown`` transition states.

The ``lifecycle_listener`` is a regular (non-lifecycle) node that subscribes to the
``lifecycle_talker``'s topics. It logs the published messages as well as the transition

> events.

Reviewer: ```suggestion
The ``lifecycle_listener`` is a regular (non-lifecycle) node that subscribes to the ``lifecycle_talker``'s topics.
It logs the published messages as well as the transition events.
```

The talker publishes only in the active state and thus the listener obviously only receives messages when the talker is in an active state.

> The ``lifecycle_service_client`` is a script calling different transitions on the ``lifecycle_talker`` through a standard ``ROS2`` node.

Reviewer: ```suggestion
The ``lifecycle_service_client`` is a script calling different transitions on the ``lifecycle_talker`` through a standard ``ROS 2`` node.
```

This represents a hypothetical external user controlling the lifecycle of nodes.

Run the demo
------------

In order to run this demo, we can either start a single launch file as in

> .. code-block:: bash

Reviewer: ```suggestion
.. code-block:: console
```
